core/src/main/java/org/infinispan/remoting/transport/jgroups/RspListFuture.java

```diff
@@ -28,12 +29,19 @@
 
    @Override
    public void futureDone(Future<RspList<Response>> future) {
+      if (timeoutFuture != null) {
+         timeoutFuture.cancel(false);
+      }
       RspList<Response> responses = request.getResults();
       complete(responses);
    }
 
    public void timeout() {
-      futureDone(null);
+      complete(request.getResults());
       request.cancel(true);
    }
+
+   public void setTimeoutFuture(Future<?> timeoutFuture) {
+      this.timeoutFuture = timeoutFuture;
```

Comment:
It seems we would have a very small edge case where if the timeout future is set right after the task completes it would still be held in memory for the full duration.  Thinking we have to check the completion state after setting the future and if it was complete then notify the timeout future manually.